Add query status event to jmx

hjafarpour opened the pull request:
### Description 
For each KSQL query we now add the query status as a JMX metric. When the status of the corresponding KafkaStreams object changes the JMX metric will be updated.
### Testing done 
Unit tests were added. Integration tests were updated.

### Reviewer checklist
- [ ] Ensure docs are updated if necessary. (eg. if a user visible feature is being added or changed).
- [ ] Ensure relevant issues are linked (description should include text like "Fixes #<issue number>")

big-andy-coates commented on ksql-engine/src/main/java/io/confluent/ksql/KsqlContext.java:
```diff
@@ -101,8 +101,9 @@ public void sql(final String sql, final Map<String, Object> overriddenProperties
       if (queryMetadata instanceof PersistentQueryMetadata) {
-        final PersistentQueryMetadata persistent = (PersistentQueryMetadata) queryMetadata;
-        persistent.getKafkaStreams().start();
+        final PersistentQueryMetadata persistentQueryMetadata
```
Did you mean to change this?

hjafarpour commented:
@big-andy-coates I like the idea of using listener to implement the query status metric. I refactored the code to use listener as you suggested.
On the metric type I was not aware of `Gauge` metric type in Kakfka. I changed the metric type to Gauge and now the value is a string (CREATED, RUNNING, RE-BALANCING, ERROR, PENDING-SHUTDOWN and NOT-RUNNING).
I also added the sum metrics for each query state so we have the counts of the number of queries in each state in a server.
And here is a screen shot of JConsole showing the query status:
![screen shot 2018-08-22 at 5 51 49 pm](https://user-images.githubusercontent.com/16784742/44801578-4a885100-ab6e-11e8-8a82-e360886d5969.png)

big-andy-coates commented on ksql-engine/src/main/java/io/confluent/ksql/internal/KsqlEngineMetrics.java:
```diff
@@ -231,4 +253,30 @@ private Sensor createSensor(final Metrics metrics, final String sensorName) {
   }
 
+
+  private Sensor configureNumActiveQueriesForGivenState(final Metrics metrics,
```
Return `Gauge<Long>` rather than `Sensor`, which is a double. A 'count' is better represented as a long...

(Same for other count metrics in this class).

hjafarpour replied:
I changed them to `Gauge<Long>` as you suggested.

big-andy-coates commented on ksql-engine/src/main/java/io/confluent/ksql/util/KsqlQueryStateUtil.java:
```diff
@@ -0,0 +1,46 @@
+    //Do nothing!
+  }
+
+  public static String getQueryStatString(final KafkaStreams.State state) {
```
Consider just dropping this class and having it in `QueryStateGauge`. It's the only thing that uses it anyway.

hjafarpour replied:
Removed it.

big-andy-coates commented on ksql-engine/src/main/java/io/confluent/ksql/internal/QueryStateListener.java:
```diff
@@ -0,0 +1,54 @@
+    this.kafkaStreams = kafkaStreams;
+    this.queryApplicationId = queryApplicationId;
+    final String metricGroupName = "ksql-queries";
+    metricName = metrics.metricName(queryApplicationId + "-query-status", metricGroupName,
```
nit: `this.metricName`.

hjafarpour replied:
Added `this`.

big-andy-coates commented on ksql-engine/src/main/java/io/confluent/ksql/internal/QueryStateListener.java:
```diff
@@ -0,0 +1,54 @@
+    final String metricGroupName = "ksql-queries";
+    metricName = metrics.metricName(queryApplicationId + "-query-status", metricGroupName,
+        "The current status of the given query.");
+    queryStateGauge = new QueryStateGauge(kafkaStreams);
```
nit: `this.queryStateGauge`.

hjafarpour replied:
Added `this`.

big-andy-coates commented on ksql-engine/src/main/java/io/confluent/ksql/util/QueryMetadata.java:
```diff
@@ -117,6 +125,9 @@ public void close() {
       log.error("Could not clean up the query with application id: {}. Query status is: {}",
                 queryApplicationId, kafkaStreams.state());
     }
+    if (queryStateListener.isPresent()) {
```
replace with `queryStateListener.ifPresent(QueryStateListener::close);`

hjafarpour replied:
Made the change.

big-andy-coates commented on ksql-engine/src/main/java/io/confluent/ksql/util/QueryMetadata.java:
```diff
@@ -170,6 +181,9 @@ public int hashCode() {
 
   public void start() {
     log.info("Starting query with application id: {}", queryApplicationId);
+    if (queryStateListener.isPresent()) {
```
replace with `queryStateListener.ifPresent(kafkaStreams::setStateListener);`

hjafarpour replied:
Made the change.

big-andy-coates commented on ksql-engine/src/test/java/io/confluent/ksql/integration/WindowingIntTest.java:
```diff
@@ -18,11 +18,13 @@
 import java.util.Map;
 import java.util.Set;
 import org.apache.kafka.clients.producer.RecordMetadata;
+import org.apache.kafka.common.metrics.Metrics;
```
nit: unused.

hjafarpour replied:
removed it.

big-andy-coates commented on ksql-engine/src/test/java/io/confluent/ksql/integration/WindowingIntTest.java:
```diff
@@ -18,11 +18,13 @@
 import org.apache.kafka.streams.kstream.TimeWindowedDeserializer;
 import org.apache.kafka.streams.kstream.Windowed;
 import org.apache.kafka.test.TestUtils;
+import org.easymock.EasyMock;
```
nit: unused.

hjafarpour replied:
removed it.

big-andy-coates commented on ksql-engine/src/main/java/io/confluent/ksql/internal/KsqlEngineMetrics.java:
```diff
@@ -66,6 +75,15 @@ public KsqlEngineMetrics(final String metricGroupPrefix, final KsqlEngine ksqlEn
     this.numIdleQueries = configureIdleQueriesSensor(metrics);
     this.messageConsumptionByQuery = configureMessageConsumptionByQuerySensor(metrics);
     this.errorRate = configureErrorRate(metrics);
+    this.createdQueriesCount = configureNumActiveQueriesForGivenState(metrics, State.CREATED);
```
Rather that having explicit lines for each state, why not just use:

```
Arrays.stream(State.values())
        .forEach(state -> configureNumActiveQueriesForGivenState(metrics, state));
```

Which means any new or changed states will automatically be added...

hjafarpour replied:
Made the change.

big-andy-coates commented on ksql-engine/src/test/java/io/confluent/ksql/internal/KsqlEngineMetricsTest.java:
```diff
@@ -90,6 +94,66 @@ public void shouldRecordNumberOfActiveQueries() {
   }
 
+  @Test
+  public void shouldRecordNumberOfQueriesInCREATEDState() {
```
Could do with some tests covering metrics for individual queries too.

hjafarpour commented on ksql-engine/src/main/java/io/confluent/ksql/KsqlContext.java:
```diff
@@ -101,8 +102,16 @@ public void sql(final String sql, final Map<String, Object> overriddenProperties
-        persistent.getKafkaStreams().start();
+        final PersistentQueryMetadata persistentQueryMetadata
+            = (PersistentQueryMetadata) queryMetadata;
+        persistentQueryMetadata.registerQueryStateListener(
```
That's a good suggestion. I made the change.

hjafarpour commented on ksql-engine/src/main/java/io/confluent/ksql/KsqlEngine.java:
```diff
@@ -526,6 +527,10 @@ public boolean terminateQuery(final QueryId queryId, final boolean closeStreams)
     return true;
   }
 
+  public Metrics getMetrics() {
```
Yes, removed it!

hjafarpour commented on ksql-engine/src/main/java/io/confluent/ksql/internal/KsqlEngineMetrics.java:
```diff
@@ -44,19 +46,26 @@
     this.metricGroupName = metricGroupPrefix + "-query-stats";
 
-    final Metrics metrics = MetricCollectors.getMetrics();
+    this.metrics = MetricCollectors.getMetrics();
```
Replaced it as you suggested.

hjafarpour commented on ksql-engine/src/main/java/io/confluent/ksql/internal/KsqlEngineMetrics.java:
```diff
@@ -83,6 +101,10 @@ public void updateMetrics() {
     recordErrorRate(MetricCollectors.currentErrorRate());
   }
 
+  public Metrics getMetrics() {
```
Removed it.

hjafarpour commented on ksql-engine/src/test/java/io/confluent/ksql/util/QueryMetadataTest.java:
```diff
@@ -0,0 +1,101 @@
+import io.confluent.ksql.planner.plan.OutputNode;
+import io.confluent.ksql.serde.DataSource;
+import io.confluent.ksql.serde.DataSource.DataSourceType;
+import java.util.ArrayList;
```
Yes, my intelliJ collapses the imports. Removed the unused ones.

hjafarpour commented on ksql-engine/src/main/java/io/confluent/ksql/internal/QueryStateGauge.java:
```diff
@@ -0,0 +1,44 @@
+    return queryState;
+  }
+
+  public String getQueryState() {
```
Removed it.

big-andy-coates commented on ksql-engine/src/main/java/io/confluent/ksql/KsqlEngine.java:
```diff
@@ -617,6 +615,8 @@ public QueryIdGenerator getQueryIdGenerator() {
         Collections.emptyMap(),
         metaStoreCopy
     );
+    engineMetrics.registerQueries(queryMetadataList);
```
Can `engineMetrics.registerQueries(queryMetadataList);` be moved into `planQueries` so its in a single place?

hjafarpour replied:
This will be updated after merging #1599.

big-andy-coates commented on ksql-engine/src/main/java/io/confluent/ksql/internal/KsqlEngineMetrics.java:
```diff
@@ -31,10 +35,13 @@
 public class KsqlEngineMetrics implements Closeable {
 
   private final List<Sensor> sensors;
+  private final List<CountMetric> countMetrics;
```
nit: can you clean up `numActiveQueries` while you're in here - this is another unused var. 

Once done, `configureNumActiveQueries` can return void.

My challenge to you it to get the file so that your IDE does not report any warnings!

hjafarpour replied:
Done! Challenge accepted :) No warning in my IDE!

big-andy-coates commented on ksql-engine/src/main/java/io/confluent/ksql/internal/KsqlEngineMetrics.java:
```diff
@@ -231,4 +261,44 @@ private Sensor createSensor(final Metrics metrics, final String sensorName) {
+  private CountMetric configureNumActiveQueriesForGivenState(final Metrics metrics,
+      final KafkaStreams.State state) {
+    final String gaugeName = ksqlServiceId + metricGroupName + "-" + state + "-queries";
+    final Gauge<Long> gauge = new Gauge<Long>() {
```
nit: replace with lambda:

```
final Gauge<Long> gauge = (metricConfig, l) -> ksqlEngine.getPersistentQueries()
 ...
```

hjafarpour replied:
Changed it to lambda.

big-andy-coates commented on ksql-engine/src/main/java/io/confluent/ksql/internal/KsqlEngineMetrics.java:
```diff
@@ -49,14 +56,21 @@
+    this(metricGroupPrefix, ksqlEngine, MetricCollectors.getMetrics());
+  }
+
+  public KsqlEngineMetrics(final String metricGroupPrefix, final KsqlEngine ksqlEngine,
```
nit: make package local - the idea is that this constructor should only be used from tests.

hjafarpour replied:
Made it package local as part of the challenge!

big-andy-coates commented on ksql-engine/src/main/java/io/confluent/ksql/internal/KsqlEngineMetrics.java:
```diff
@@ -231,4 +261,44 @@ private Sensor createSensor(final Metrics metrics, final String sensorName) {
+    private final MetricName metricName;
+
+    CountMetric(final MetricName metricName, final Gauge<Long> count) {
+      this.metricName = metricName;
```
nit: null check on params.

hjafarpour replied:
Added null check.

big-andy-coates commented on ksql-engine/src/main/java/io/confluent/ksql/internal/KsqlEngineMetrics.java:
```diff
@@ -231,4 +261,44 @@ private Sensor createSensor(final Metrics metrics, final String sensorName) {
+    return countMetric;
+  }
+
+  private static class CountMetric {
```
nit: move out into own class, rather than nested.

hjafarpour replied:
This is only being used here. Any specific reason to make it it's own class?

big-andy-coates replied:
To reduce the complexity of this class.

big-andy-coates commented on checkstyle/suppressions.xml:
```diff
@@ -18,7 +18,7 @@
               checks="ClassDataAbstractionCouplingCheck"/>
     <suppress
-            files="(QueryEngine|LogicalPlanner|FunctionRegistry|KsqlEngine|SchemaKStream|Analyzer|AggregateNode).java" checks="ClassDataAbstractionCouplingCheck"/>
+            files="(QueryEngine|LogicalPlanner|FunctionRegistry|KsqlEngine|SchemaKStream|Analyzer|AggregateNode|KsqlEngineMetrics).java" checks="ClassDataAbstractionCouplingCheck"/>
```
I think its hard to justify disabling this check on KsqlEngineMetrics. It's not some class wiring together a load of others. Much better to look at how you can refactor the class. Otherwise, what's the point in the checkstyle rule if we're just going to disable it each time we hit it?

For example, if you change the class to use a lambda for the gauge, as I've suggested, you no longer need this suppression.

hjafarpour replied:
Thanks for the suggestion Andy! Using lambda, there is no need for this suppression. Removed it!
